// @aws-sdk/client-chime-sdk-meetings/dist-es/commands/CreateMeetingWithAttendeesCommand.js
import { getEndpointPlugin } from "@smithy/middleware-endpoint";
import { getSerdePlugin } from "@smithy/middleware-serde";
import { Command as $Command } from "@smithy/smithy-client";
import { SMITHY_CONTEXT_KEY, } from "@smithy/types";
import { CreateMeetingWithAttendeesRequestFilterSensitiveLog, CreateMeetingWithAttendeesResponseFilterSensitiveLog, } from "../models/models_0";
import { de_CreateMeetingWithAttendeesCommand, se_CreateMeetingWithAttendeesCommand } from "../protocols/Aws_restJson1";
export { $Command };
export class CreateMeetingWithAttendeesCommand extends $Command {
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
        };
    }
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, CreateMeetingWithAttendeesCommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "ChimeSDKMeetingsClient";
        const commandName = "CreateMeetingWithAttendeesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: CreateMeetingWithAttendeesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: CreateMeetingWithAttendeesResponseFilterSensitiveLog,
            [SMITHY_CONTEXT_KEY]: {
                service: "ChimeMeetingsSDKService",
                operation: "CreateMeetingWithAttendees",
            },
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return se_CreateMeetingWithAttendeesCommand(input, context);
    }
    deserialize(output, context) {
        return de_CreateMeetingWithAttendeesCommand(output, context);
    }
}
